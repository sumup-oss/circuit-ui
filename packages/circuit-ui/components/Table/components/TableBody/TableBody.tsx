/**
 * Copyright 2019, SumUp Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { mapRowProps, mapCellProps } from '../../utils';
import { Row } from '../../types';
import TableRow from '../TableRow';
import TableHeader from '../TableHeader';
import TableCell from '../TableCell';

type TableBodyProps = {
  /**
   * An array of rows or an object with children containing an array of cells
   * for the table.
   */
  rows?: Row[];
  /**
   * Enables/disables sticky columns on mobile.
   */
  rowHeaders?: boolean;
  /**
   * The current hovered sort cell index.
   */
  sortHover?: number;
  /**
   * Adds condensed styles to the table.
   */
  condensed?: boolean;
  /**
   * A function to call when the row is clicked.
   */
  onRowClick?: (rowIndex: number) => void;
};

/**
 * TableBody for the Table component. The Table handles rendering it.
 */
const TableBody = ({
  rows = [],
  condensed,
  rowHeaders = false,
  sortHover,
  onRowClick,
}: TableBodyProps): JSX.Element => (
  <tbody>
    {rows.map((row, rowIndex) => {
      const { cells, ...props } = mapRowProps(row);
      return (
        <TableRow
          key={`table-row-${rowIndex}`}
          onClick={onRowClick ? () => onRowClick(rowIndex) : undefined}
          {...props}
        >
          {cells.map((cell, cellIndex) =>
            rowHeaders && cellIndex === 0 ? (
              <TableHeader
                key={`table-cell-${rowIndex}-${cellIndex}`}
                fixed
                condensed={condensed}
                scope="row"
                isHovered={sortHover === cellIndex}
                sortParams={{ sortable: false }}
                {...mapCellProps(cell)}
              />
            ) : (
              <TableCell
                key={`table-cell-${rowIndex}-${cellIndex}`}
                condensed={condensed}
                isHovered={sortHover === cellIndex}
                {...mapCellProps(cell)}
              />
            ),
          )}
        </TableRow>
      );
    })}
  </tbody>
);

export default TableBody;
