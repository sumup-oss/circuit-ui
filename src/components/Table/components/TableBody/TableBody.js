/**
 * Copyright 2019, SumUp Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React, { Fragment } from 'react';
import PropTypes from 'prop-types';

import {
  mapRowProps,
  mapCellProps,
  getCellChildren,
  RowPropType
} from '../../utils';
import { TR_KEY_PREFIX, TD_KEY_PREFIX } from '../../constants';
import TableRow from '../TableRow';
import TableHeader from '../TableHeader';
import TableCell from '../TableCell';

const getRowKey = index => `${TR_KEY_PREFIX}-${index}`;
const getCellKey = (rowIndex, cellIndex) =>
  `${TD_KEY_PREFIX}-${rowIndex}-${cellIndex}`;

const TableBody = ({ rows, condensed, rowHeaders, sortHover, onRowClick }) => (
  <tbody>
    {rows.map((row, rowIndex) => {
      const { cells, ...props } = mapRowProps(row);
      return (
        <TableRow
          key={getRowKey(rowIndex)}
          data-testid="table-row"
          onClick={onRowClick ? () => onRowClick(rowIndex) : null}
          {...props}
        >
          {cells.map((cell, cellIndex) =>
            rowHeaders && cellIndex === 0 ? (
              <Fragment key={getCellKey(rowIndex, cellIndex)}>
                <TableHeader
                  fixed
                  condensed={condensed}
                  scope={TableHeader.ROW}
                  isHovered={sortHover === cellIndex}
                  {...mapCellProps(cell)}
                />
                <TableCell
                  role="presentation"
                  condensed={condensed}
                  aria-hidden="true"
                >
                  {getCellChildren(cell)}
                </TableCell>
              </Fragment>
            ) : (
              <TableCell
                key={getCellKey(rowIndex, cellIndex)}
                condensed={condensed}
                isHovered={sortHover === cellIndex}
                {...mapCellProps(cell)}
              />
            )
          )}
        </TableRow>
      );
    })}
  </tbody>
);

/**
 * @private TableHead for the Table component. The Table handles rendering it
 */
TableBody.propTypes = {
  /**
   *(An array of rows or object with children) containing cells for the table. The Cell can be a string
   * or an object with options described on TableHeader component
   */
  rows: PropTypes.arrayOf(
    PropTypes.oneOfType([
      PropTypes.shape({ cells: PropTypes.arrayOf(RowPropType) }),
      PropTypes.arrayOf(RowPropType)
    ])
  ),
  /**
   * Enables/disables sticky columns on mobile
   */
  rowHeaders: PropTypes.bool,
  /**
   * @private The current hovered sort cell index
   * Handled internally
   */
  sortHover: PropTypes.number,
  /**
   * @private Adds condensed styles to the table.
   * Handled internally
   */
  condensed: PropTypes.bool,
  /**
   * onClick handler
   */
  onRowClick: PropTypes.func
};

TableBody.defaultProps = {
  rows: [],
  rowHeaders: false,
  sortHover: null,
  onRowClick: null
};

export default TableBody;
