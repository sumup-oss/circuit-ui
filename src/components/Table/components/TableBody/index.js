import React, { Fragment } from 'react';
import PropTypes from 'prop-types';

import TableRow from '../TableRow';
import TableHeader from '../TableHeader';
import TableCell from '../TableCell';
import { mapProps, getChildren, RowPropType } from '../../utils';
import { TR_KEY_PREFIX, TD_KEY_PREFIX } from '../../constants';

const getRowKey = index => `${TR_KEY_PREFIX}-${index}`;
const getCellKey = (rowIndex, cellIndex) =>
  `${TD_KEY_PREFIX}-${rowIndex}-${cellIndex}`;

const TableBody = ({ rows, rowHeaders, sortHover }) => (
  <tbody>
    {rows.map((row, rowIndex) => (
      // eslint-disable-next-line react/no-array-index-key
      <TableRow key={getRowKey(rowIndex)}>
        {row.map(
          (cell, cellIndex) =>
            rowHeaders && cellIndex === 0 ? (
              // eslint-disable-next-line react/no-array-index-key
              <Fragment key={getCellKey(rowIndex, cellIndex)}>
                <TableHeader
                  fixed
                  scope={TableHeader.ROW}
                  isHovered={sortHover === cellIndex}
                  {...mapProps(cell)}
                />
                <TableCell role="presentation" aria-hidden="true">
                  {getChildren(cell)}
                </TableCell>
              </Fragment>
            ) : (
              <TableCell
                // eslint-disable-next-line react/no-array-index-key
                key={getCellKey(rowIndex, cellIndex)}
                isHovered={sortHover === cellIndex}
                {...mapProps(cell)}
              />
            )
        )}
      </TableRow>
    ))}
  </tbody>
);

/**
 * [PRIVATE] TableHead for the Table component. The Table handles rendering it
 */
TableBody.propTypes = {
  /**
   * An array of rows with cells for the table. The Cell can be a string
   * or an object with options described on TableHeader component
   */
  rows: PropTypes.arrayOf(PropTypes.arrayOf(RowPropType)),
  /**
   * Enables/disables sticky columns on mobile
   */
  rowHeaders: PropTypes.bool,
  /**
   * [PRIVATE] The current hovered sort cell index
   * Handled internally
   */
  sortHover: PropTypes.number
};

TableBody.defaultProps = {
  rows: [],
  rowHeaders: false,
  sortHover: null
};

export default TableBody;
