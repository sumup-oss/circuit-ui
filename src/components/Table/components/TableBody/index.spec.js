import React from 'react';

import TableBody from '.';

const fixtureRows = [['Foo', 'Bar']];

describe('TableBody', () => {
  describe('Style tests', () => {
    it('should render with default styles', () => {
      const actual = create(<TableBody rows={fixtureRows} />);
      expect(actual).toMatchSnapshot();
    });

    it('should render with fixed Header styles', () => {
      const actual = create(<TableBody rowHeaders rows={fixtureRows} />);
      expect(actual).toMatchSnapshot();
    });
  });

  describe('rowHeaders', () => {
    describe('no rowHeaders', () => {
      it('should render a TableCell on the first cellIndex of each row', () => {
        const wrapper = shallow(<TableBody rows={fixtureRows} />);

        expect(
          wrapper
            .find('TableRow')
            .at(0)
            .find('TableCell')
            .at(0).length
        ).toBeTruthy();
      });
    });

    it('should add fixed to first TableHeader', () => {
      const wrapper = shallow(<TableBody rowHeaders rows={fixtureRows} />);

      expect(
        wrapper
          .find('TableRow')
          .at(0)
          .find('TableHeader')
          .at(0)
          .props().fixed
      ).toBeTruthy();
    });

    describe('presentational TableCell', () => {
      it('should add it to the first header', () => {
        const wrapper = shallow(<TableBody rowHeaders rows={fixtureRows} />);

        expect(
          wrapper
            .find('TableRow')
            .at(0)
            .find('TableCell')
            .at(0).length
        ).toBeTruthy();
      });
    });
  });

  describe('sortHover', () => {
    describe('Cell index not relative to sortHover', () => {
      it('should not add isHovered to it', () => {
        const sortHover = 1;
        const wrapper = shallow(
          <TableBody sortHover={sortHover} rows={fixtureRows} />
        );

        expect(
          wrapper
            .find('TableRow')
            .at(0)
            .find('TableCell')
            .at(0)
            .props().isHovered
        ).toBeFalsy();
      });
    });
    describe('relative Cell index to sortHover', () => {
      it('should add isHovered to it', () => {
        const sortHover = 0;
        const wrapper = shallow(
          <TableBody sortHover={sortHover} rows={fixtureRows} />
        );

        expect(
          wrapper
            .find('TableRow')
            .at(0)
            .find('TableCell')
            .at(0)
            .props().isHovered
        ).toBeTruthy();
      });
    });
  });

  describe('Accessibility tests', () => {
    it('should meet accessibility guidelines', async () => {
      const wrapper = renderToHtml(<TableBody rowHeader rows={fixtureRows} />);
      const actual = await axe(wrapper);
      expect(actual).toHaveNoViolations();
    });
  });
});
