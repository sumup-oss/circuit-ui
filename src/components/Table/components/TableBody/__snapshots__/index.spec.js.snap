// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TableBody Style tests should render with default styles 1`] = `
.circuit-4 {
  vertical-align: middle;
}

.circuit-0 {
  background-color: #FFFFFF;
  border-bottom: 1px solid #D8DDE1;
  padding: 24px;
  text-align: left;
  -webkit-transition: background-color 200ms ease-in-out;
  transition: background-color 200ms ease-in-out;
  vertical-align: middle;
  white-space: nowrap;
}

<tbody>
  <tr
    className="circuit-4 circuit-5"
  >
    <td
      className="circuit-0 circuit-1"
    >
      Foo
    </td>
    <td
      className="circuit-0 circuit-1"
    >
      Bar
    </td>
  </tr>
</tbody>
`;

exports[`TableBody Style tests should render with fixed Header styles 1`] = `
.circuit-6 {
  vertical-align: middle;
}

.circuit-4 {
  background-color: #FFFFFF;
  border-bottom: 1px solid #D8DDE1;
  padding: 24px;
  text-align: left;
  -webkit-transition: background-color 200ms ease-in-out;
  transition: background-color 200ms ease-in-out;
  vertical-align: middle;
  white-space: nowrap;
}

.circuit-0 {
  background-color: #FFFFFF;
  border-bottom: 1px solid #D8DDE1;
  padding: 24px;
  text-align: left;
  -webkit-transition: background-color 200ms ease-in-out,color 200ms ease-in-out;
  transition: background-color 200ms ease-in-out,color 200ms ease-in-out;
  white-space: nowrap;
}

@media (max-width:767px) {
  .circuit-0 {
    left: 0;
    top: auto;
    position: absolute;
    width: 145px;
    white-space: unset;
  }
}

.circuit-2 {
  background-color: #FFFFFF;
  border-bottom: 1px solid #D8DDE1;
  padding: 24px;
  text-align: left;
  -webkit-transition: background-color 200ms ease-in-out;
  transition: background-color 200ms ease-in-out;
  vertical-align: middle;
  white-space: nowrap;
  display: none;
}

@media (max-width:767px) {
  .circuit-2 {
    display: table-cell;
    min-width: 145px;
    white-space: unset;
    width: 145px;
  }
}

<tbody>
  <tr
    className="circuit-6 circuit-7"
  >
    <th
      aria-sort={null}
      className="circuit-0 circuit-1"
      scope="row"
    >
      Foo
    </th>
    <td
      aria-hidden="true"
      className="circuit-2 circuit-3"
      role="presentation"
    >
      Foo
    </td>
    <td
      className="circuit-4 circuit-3"
    >
      Bar
    </td>
  </tr>
</tbody>
`;
